### PR TITLE
Fix plugin config per environment doesn't apply

### LINKED ISSUE #11774: Global config getter does not return per-environment config.
### Describe the bug

The config getter does not return config that is configured in `/env/{environment}/*.js` as per these [environment configurations docs](https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/configurations/optional/environment.html#environment-configurations).

This for example will result in not being able to load an different upload provider as the upload plugin is using `strapi.config.get('plugin.upload', {})`:
https://github.com/strapi/strapi/blob/11f3519710b77042fc58b787acd46288930b510e/packages/core/upload/server/bootstrap.js#L9

### Steps to reproduce the behavior

1. `yarn create strapi-app --quickstart`
2. Create `./config/env/development/plugins.js` with
  ```js
  module.exports = {
    upload: {
      enabled: true,
      config: {
        provider: 'local',
        providerOptions: {
          sizeLimit: 2000000,
        },
      },
    },
  };
  ```
3. Start interactive CLI `yarn strapi console`
4. Try and see differences between `strapi.config.plugins` and `strapi.config.get('plugin.upload', {})`

### Expected behavior

Config getter should respect the per-environment configurations

### Screenshots

image

### System

- Node.js version: `14.16.1`
- NPM version: `7.19.1`
- Yarn version: `1.22.10`
- Strapi version: `4.0.0`
- Database: SQLite (Quickstart)
- Operating system: Ubuntu 20.04

### PR DESCRIPTION
**Pull request by iicdii**
Signed-off-by: harimkims <harimkims@gmail.com>

### What does it do?

I added logic to find `config/env/<process.env.NODE_ENV>/plugins.js` first when it loads plugin config.
**As-is**: `config/plugins.js` -> default plugin config
**To-be**: `config/env/${process.env.NODE_ENV}/plugins.js` -> `config/plugins.js` -> default plugin config

### Why is it needed?

To support [Environment configurations](https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/configurations/optional/environment.html#environment-configurations)

### How to test it?

1. Create `config/env/development/plugins.js`
2. Copy and taste config below
```js
module.exports = () => ({
  upload: {
    enabled: true,
    config: {
      provider: 'local',
      providerOptions: {
        sizeLimit: 3000000,
      },
    },
  },
});
```
3. Start interactive CLI `yarn strapi console`
4. Try `strapi.config.get('plugin.upload', {})` and see difference
5. Create `config/plugins.js`
6. Repeat Step 2~4 but change `sizeLimit` to 2000000
7. Delete `config/plugins.js`
8. Repeat Step 3~4 (default `sizeLimit` is 1000000 though)

### Related issue(s)/PR(s)

Closes #11774

### CODE REVIEW

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
Ideally, we would like this to work the same way as other configurations, 

Loading the main plugins.js file and merging the env/x/plugins.js file with it instead of switching. This would make the behavior more consistent :)

**Comment by iicdii** [comment]
I applied your feedback. What do you think?

File: packages/core/strapi/lib/core/loaders/plugins/index.js [hunk 1]
```diff
@@ -65,11 +65,29 @@ const formatContentTypes = plugins => {
   }
 };
 
+const getUserPluginsConfig = async () => {
+  const globalUserConfigPath = join(strapi.dirs.config, 'plugins.js');
+  const currentEnvUserConfigPath = join(
+    strapi.dirs.config,
+    `env/${process.env.NODE_ENV}/plugins.js`
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/loaders/plugins/index.js:72]: to simplify you can do 
```js
join(strapi.dirs.config, 'env', process.env.NODE_ENV, 'plugins.js')
```

**Review by alexandrebodin — COMMENTED** [review]
Just a simplification in the code but that looks quite good. I'll test it

**Review by alexandrebodin — COMMENTED** [review]
Something that will not be working is disabling a plugin in an env. The getEnablePlugin code loads the config again. I think this deserves some refactoring :)

**Comment by iicdii** [comment]
I changed 2 parts.
1. in `get-enabled-plugins.js` function, DeclaredPlugins didn't look for config exist in `config/env/${env}/plugins.js`.. I fixed it. Now it cares about user's config by environment.
2. I found that the `getEnablePlugin` code enables every plugin whether user configured it `true` or `false` or whatever. I fixed it so now user's `enabled: false` flag is applied to config. The problem was Lodash `defaultsDeep` function ignores the property if it has `enabled` key already. I kept `defaultsDeep` function, but I applied user's enabled flag after that.

**Comment by alexandrebodin** [comment]
> I reproduced the problem you said. I changed 2 parts.
> 
> 1. I fixed `declaredPlugins` in `get-enabled-plugins.js` function didn't look for config exist in `config/env/${env}/plugins.js`. Now it cares about user's config by environment.
> 2. I found that the `getEnablePlugin` code enables every plugin whether user configured it `true` or `false` or whatever. I fixed it so now user's `enabled: false` flag is applied to config. The problem was Lodash `defaultsDeep` function ignores the property if it has `enabled` key already. I kept `defaultsDeep` function and I applied user's enabled flag after that.

I never had an issue with disable a plugin I'm not sure this is accurate 🤔

**Comment by iicdii** [comment]
You're right.

> I found that the getEnablePlugin code enables every plugin whether user configured it true or false or whatever. I fixed it so now user's enabled: false flag is applied to config. The problem was Lodash defaultsDeep function ignores the property if it has enabled key already. I kept defaultsDeep function and I applied user's enabled flag after that.

It was wrong. I reverted this fix to the original code. However, the first modification - `get-enabled-plugins.js` function didn't look for config exist in `config/env/${env}/plugins.js`  - is still valid.

File: packages/core/strapi/lib/core/loaders/plugins/get-enabled-plugins.js [hunk 2]
```diff
@@ -106,6 +103,7 @@ const getEnabledPlugins = async strapi => {
 
   const enabledPlugins = pipe(
     defaultsDeep(declaredPlugins),
+    pickBy((pValue, pName) => _.get(declaredPlugins, [pName, 'enabled'], true)),
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/loaders/plugins/get-enabled-plugins.js:106]: is this necessary? the last pickBy in the list should be doing the job

Comment (iicdii) [reply packages/core/strapi/lib/core/loaders/plugins/get-enabled-plugins.js]: No It's not, I just pushed commit

**Review by alexandrebodin — APPROVED** [review]
This works like a charm. Thanks you for this great contribution !

**Pull request merged**